File: monitoring-client/src/file_sync_manager.py
"""File synchronization manager for OTA file transfers."""

import os
import time
import threading
import requests
from pathlib import Path
from typing import List, Dict, Optional
from queue import Queue
from .logger import get_logger
from .config import Config, retrieve_client_id

logger = get_logger()


class FileSyncManager:
    """Manages file synchronization from server to client."""
    
    def __init__(self, config: Config, employee_name: str):
        """
        Initialize file sync manager.
        
        Args:
            config: Configuration object
            employee_name: Name of the employee
        """
        self.config = config
        self.employee_name = employee_name
        self.client_id = retrieve_client_id()  # Get or generate unique client ID
        self.download_path = self._get_download_path()
        self.poll_interval = self._get_poll_interval()
        self.max_parallel = 5
        self.running = False
        self.download_queue: Queue = Queue()
        self.active_downloads: List[threading.Thread] = []
        self.downloaded_files: set = set()  # Set of file IDs
        self.file_id_to_name: Dict[str, str] = {}  # Map file ID to filename for deletion
        self.check_counter: int = 0  # Counter for logging
        
        # Ensure download directory exists
        self.download_path.mkdir(parents=True, exist_ok=True)
        
        logger.info(f"📁 File sync manager initialized. Download path: {self.download_path}")
        logger.info(f"🆔 Client ID: {self.client_id}")
        logger.info(f"⏱️  File sync will check every {self.poll_interval} seconds")
    
    def _get_download_path(self) -> Path:
        """Get the download path from environment or use default."""
        path_str = os.getenv('FILE_DOWNLOAD_PATH', 'C:\\Downloads\\CompanyFiles')
        return Path(path_str)
    
    def _get_poll_interval(self) -> int:
        """
        Get the polling interval in seconds.
        NOTE: This is no longer used since file sync now happens on every app usage poll.
        Kept for backward compatibility with config.
        """
        try:
            return int(os.getenv('FILE_SYNC_INTERVAL', '10'))
        except ValueError:
            return 10
    
    def start(self):
        """Start the file sync manager."""
        if self.running:
            logger.warning("⚠️  File sync manager already running")
            return
        
        self.running = True
        
        print(f"✅ FILE SYNC: Started (checks on every app usage poll)")
        logger.info(f"✅ File sync manager started")
        logger.info(f"📁 Download path: {self.download_path}")
        logger.info(f"👤 Employee: {self.employee_name}")
        logger.info(f"⏱️  Will check on every app usage poll (no separate interval)")
        
        # Do an immediate check
        logger.info("🔍 Performing initial file check...")
        print("🔍 FILE SYNC: Performing initial check...")
        try:
            self._check_pending_files()
            self._process_download_queue()
        except Exception as e:
            logger.error(f"❌ Error in initial file check: {e}", exc_info=True)
            print(f"❌ FILE SYNC: Error in initial check: {e}")
    
    def stop(self):
        """Stop the file sync manager."""
        self.running = False
        print("🛑 FILE SYNC: Stopped")
        logger.info("🛑 File sync manager stopped")
    
    def update(self):
        """
        Update file sync - check for files on every call.
        This is called from the app usage tracker's polling loop (every 10 seconds).
        """
        if not self.running:
            return
        
        self.check_counter += 1
        
        # Check on every update (no separate interval)
        try:
            self._check_pending_files()
            self._process_download_queue()
        except Exception as e:
            logger.error(f"❌ Error in file sync update: {e}", exc_info=True)
            print(f"❌ FILE SYNC: Error - {e}")
    
    def check_and_download(self):
        """
        Force an immediate check for pending files and process downloads.
        This can be called manually or from WebSocket events.
        """
        if not self.running:
            return
        
        print("🔍 FILE SYNC: Force check triggered")
        logger.info("🔍 File sync: Force check triggered")
        
        try:
            self._check_pending_files()
            self._process_download_queue()
        except Exception as e:
            logger.error(f"❌ Error in check_and_download: {e}", exc_info=True)
            print(f"❌ FILE SYNC: Error - {e}")
    
    def _check_pending_files(self):
        """Check server for pending files and sync deletions."""
        try:
            # Extract base URL (remove /api/monitoring/data if present)
            base_url = self.config.server_url.replace('/api/monitoring/data', '')
            
            # First, check for pending files to download
            pending_url = f"{base_url}/api/files/pending/{self.employee_name}?client_id={self.client_id}"
            
            print(f"🌐 FILE SYNC: Checking {pending_url}")
            logger.info(f"🌐 Checking pending files at: {pending_url}")
            logger.debug(f"👤 Employee name: {self.employee_name}")
            logger.debug(f"🆔 Client ID: {self.client_id}")
            
            headers = {
                'Authorization': f'Bearer {self.config.auth_token}'
            }
            
            response = requests.get(pending_url, headers=headers, timeout=10)
            
            logger.debug(f"📊 Response status: {response.status_code}")
            
            if response.status_code == 200:
                data = response.json()
                files = data.get('files', [])
                
                if len(files) > 0:
                    print(f"📦 FILE SYNC: Found {len(files)} pending file(s)!")
                    logger.info(f"📦 Found {len(files)} pending file(s)")
                # Only log "no pending files" occasionally to reduce noise
                elif self.check_counter % 6 == 1:  # Every ~60 seconds (6 polls * 10s)
                    print(f"✅ FILE SYNC: No pending files (check #{self.check_counter})")
                    logger.info(f"✅ No pending files")
                
                for file_info in files:
                    file_id = file_info['id']
                    filename = file_info['filename']
                    filesize = file_info.get('fileSize', 0)
                    
                    print(f"📄 FILE SYNC: {filename} ({filesize} bytes, ID: {file_id})")
                    logger.info(f"📄 File: {filename} (ID: {file_id}, Size: {filesize})")
                    
                    if file_id not in self.downloaded_files:
                        self.download_queue.put(file_info)
                        print(f"✅ FILE SYNC: Queued for download: {filename}")
                        logger.info(f"✅ Queued file for download: {filename}")
                    else:
                        print(f"⏭️  FILE SYNC: Already downloaded: {filename}")
                        logger.debug(f"⏭️  File already downloaded: {filename}")
            else:
                print(f"⚠️  FILE SYNC: HTTP {response.status_code}")
                logger.warning(f"⚠️  Failed to check pending files: HTTP {response.status_code}")
                logger.debug(f"Response body: {response.text[:200]}")
            
            # Now check for deleted files by getting list of active files
            print("🔍 FILE SYNC: Checking for deleted files...")
            logger.info("🔍 Checking for deleted files...")
            self._sync_deletions(base_url, headers)
        
        except Exception as e:
            print(f"❌ FILE SYNC: Error checking files - {e}")
            logger.error(f"❌ Error checking pending files: {e}", exc_info=True)
    
    def _sync_deletions(self, base_url: str, headers: Dict):
        """
        Check for deleted files and remove them from local storage.
        
        Args:
            base_url: Base URL of the server
            headers: HTTP headers with auth token
        """
        try:
            # Get list of all active files from server
            active_url = f"{base_url}/api/files/active/{self.employee_name}"
            
            print(f"🌐 FILE SYNC: Checking active files at {active_url}")
            logger.info(f"🔍 Checking for deleted files at: {active_url}")
            
            response = requests.get(active_url, headers=headers, timeout=10)
            
            print(f"📊 FILE SYNC: Active files response: {response.status_code}")
            
            if response.status_code == 200:
                data = response.json()
                active_files = data.get('files', [])
                
                # Build set of filenames that should exist
                server_filenames = {f['filename'] for f in active_files}
                
                print(f"📋 FILE SYNC: Server has {len(server_filenames)} active file(s): {server_filenames}")
                logger.info(f"📋 Server has {len(server_filenames)} active file(s)")
                
                # Check local files
                if self.download_path.exists():
                    local_files = list(self.download_path.iterdir())
                    local_filenames = [f.name for f in local_files if f.is_file()]
                    
                    print(f"📂 FILE SYNC: Local has {len(local_filenames)} file(s): {local_filenames}")
                    
                    for local_file in local_files:
                        if local_file.is_file():
                            filename = local_file.name
                            
                            # If file exists locally but not on server, delete it
                            if filename not in server_filenames:
                                try:
                                    local_file.unlink()
                                    print(f"🗑️  FILE SYNC: Deleted orphaned file: {filename}")
                                    logger.info(f"🗑️  Deleted orphaned file: {filename}")
                                    
                                    # Remove from tracking
                                    file_id_to_remove = None
                                    for fid, fname in self.file_id_to_name.items():
                                        if fname == filename:
                                            file_id_to_remove = fid
                                            break
                                    
                                    if file_id_to_remove:
                                        if file_id_to_remove in self.downloaded_files:
                                            self.downloaded_files.remove(file_id_to_remove)
                                        if file_id_to_remove in self.file_id_to_name:
                                            del self.file_id_to_name[file_id_to_remove]
                                
                                except Exception as e:
                                    print(f"❌ FILE SYNC: Error deleting {filename}: {e}")
                                    logger.error(f"Error deleting orphaned file {filename}: {e}")
                else:
                    print("📂 FILE SYNC: Download directory doesn't exist")
            else:
                print(f"⚠️  FILE SYNC: Could not check for deletions: HTTP {response.status_code}")
                logger.warning(f"Could not check for deletions: HTTP {response.status_code}")
        
        except Exception as e:
            print(f"❌ FILE SYNC: Error syncing deletions: {e}")
            logger.error(f"Error syncing deletions: {e}", exc_info=True)
    
    def _process_download_queue(self):
        """Process the download queue with parallel downloads."""
        # Clean up completed threads
        self.active_downloads = [t for t in self.active_downloads if t.is_alive()]
        
        queue_size = self.download_queue.qsize()
        active_count = len(self.active_downloads)
        
        if queue_size > 0 or active_count > 0:
            print(f"📥 FILE SYNC: Queue={queue_size}, Active={active_count}")
            logger.info(f"📥 Download queue: {queue_size} pending, {active_count} active")
        
        # Start new downloads up to max parallel limit
        while len(self.active_downloads) < self.max_parallel and not self.download_queue.empty():
            file_info = self.download_queue.get()
            print(f"🚀 FILE SYNC: Starting download thread for: {file_info['filename']}")
            logger.info(f"🚀 Starting download thread for: {file_info['filename']}")
            thread = threading.Thread(
                target=self._download_file,
                args=(file_info,),
                daemon=True
            )
            thread.start()
            self.active_downloads.append(thread)
    
    def _download_file(self, file_info: Dict):
        """
        Download a file from the server.
        
        Args:
            file_info: Dictionary containing file information
        """
        file_id = file_info['id']
        filename = file_info['filename']
        
        try:
            print(f"⬇️  FILE SYNC: Downloading {filename}...")
            logger.info(f"⬇️  Starting download: {filename}")
            logger.debug(f"File ID: {file_id}, Size: {file_info.get('fileSize', 'unknown')}")
            
            # Update status to downloading
            self._update_status(file_id, 'downloading')
            
            # Extract base URL (remove /api/monitoring/data if present)
            base_url = self.config.server_url.replace('/api/monitoring/data', '')
            
            # Download file
            url = f"{base_url}/api/files/{file_id}/download"
            logger.debug(f"🌐 Download URL: {url}")
            
            headers = {
                'Authorization': f'Bearer {self.config.auth_token}'
            }
            params = {
                'employeeName': self.employee_name
            }
            
            logger.debug(f"📡 Sending download request...")
            response = requests.get(
                url,
                headers=headers,
                params=params,
                stream=True,
                timeout=300  # 5 minutes timeout for large files
            )
            
            logger.debug(f"📊 Download response status: {response.status_code}")
            
            if response.status_code == 200:
                # Save file
                file_path = self.download_path / filename
                print(f"💾 FILE SYNC: Saving to {file_path}")
                logger.info(f"💾 Saving file to: {file_path}")
                
                bytes_downloaded = 0
                with open(file_path, 'wb') as f:
                    for chunk in response.iter_content(chunk_size=8192):
                        if chunk:
                            f.write(chunk)
                            bytes_downloaded += len(chunk)
                
                print(f"✅ FILE SYNC: Downloaded {filename} ({bytes_downloaded} bytes)")
                logger.info(f"✅ Download completed: {filename} ({bytes_downloaded} bytes)")
                logger.debug(f"File saved at: {file_path}")
                
                # Update status to completed
                self._update_status(file_id, 'completed')
                self.downloaded_files.add(file_id)
                
                # Track file ID to filename mapping for deletion
                self.file_id_to_name[file_id] = filename
                
            else:
                error_msg = f"Download failed with status {response.status_code}"
                print(f"❌ FILE SYNC: {error_msg} - {filename}")
                logger.error(f"❌ {error_msg}: {filename}")
                logger.debug(f"Response body: {response.text[:200]}")
                self._update_status(file_id, 'failed', error_msg)
        
        except Exception as e:
            error_msg = str(e)
            print(f"❌ FILE SYNC: Error downloading {filename} - {error_msg}")
            logger.error(f"❌ Error downloading file {filename}: {error_msg}", exc_info=True)
            self._update_status(file_id, 'failed', error_msg)
    
    def _update_status(self, file_id: str, status: str, error: Optional[str] = None):
        """
        Update download status on server.
        
        Args:
            file_id: File transfer ID
            status: Status to update (downloading, completed, failed)
            error: Optional error message
        """
        try:
            # Extract base URL (remove /api/monitoring/data if present)
            base_url = self.config.server_url.replace('/api/monitoring/data', '')
            
            url = f"{base_url}/api/files/{file_id}/status"
            headers = {
                'Authorization': f'Bearer {self.config.auth_token}',
                'Content-Type': 'application/json'
            }
            data = {
                'employeeName': self.employee_name,
                'status': status
            }
            
            if error:
                data['error'] = error
            
            response = requests.post(url, headers=headers, json=data, timeout=10)
            
            if response.status_code != 200:
                logger.warning(f"Failed to update status: {response.status_code}")
        
        except Exception as e:
            logger.error(f"Error updating status: {e}")
    
    def handle_file_uploaded(self, payload: Dict):
        """
        Handle file uploaded WebSocket event.
        
        Args:
            payload: Event payload containing file information
        """
        filename = payload.get('filename', 'unknown')
        print(f"📢 FILE SYNC: New file uploaded - {filename}")
        logger.info(f"📢 Received file upload notification: {filename}")
        
        # Immediately check for pending files
        print("🔍 FILE SYNC: Checking for new files...")
        self.check_and_download()
    
    def handle_file_deleted(self, payload: Dict):
        """
        Handle file deleted WebSocket event.
        Deletes the local file from the download directory.
        
        Args:
            payload: Event payload containing file ID and filename
        """
        file_id = payload.get('fileId')
        
        print(f"🗑️  FILE SYNC: Received deletion request for file ID: {file_id}")
        logger.info(f"🗑️  Received file deletion notification: {file_id}")
        
        # Try to get filename from mapping
        filename = self.file_id_to_name.get(file_id)
        
        if not filename:
            # If not in mapping, try to get from payload
            filename = payload.get('filename')
            print(f"⚠️  FILE SYNC: Filename not in cache, using payload: {filename}")
            logger.warning(f"Filename not in cache for file ID {file_id}, using payload")
        
        if filename:
            file_path = self.download_path / filename
            
            try:
                if file_path.exists():
                    file_path.unlink()
                    print(f"✅ FILE SYNC: Deleted local file: {filename}")
                    logger.info(f"✅ Deleted local file: {file_path}")
                else:
                    print(f"⚠️  FILE SYNC: File not found locally: {filename}")
                    logger.warning(f"File not found locally: {file_path}")
            except Exception as e:
                print(f"❌ FILE SYNC: Error deleting file {filename}: {e}")
                logger.error(f"❌ Error deleting file {filename}: {e}", exc_info=True)
        else:
            print(f"❌ FILE SYNC: Cannot delete - filename unknown for ID: {file_id}")
            logger.error(f"Cannot delete file - filename unknown for ID: {file_id}")
        
        # Remove from tracking sets
        if file_id in self.downloaded_files:
            self.downloaded_files.remove(file_id)
            print(f"🗑️  FILE SYNC: Removed {file_id} from downloaded files tracking")
        
        if file_id in self.file_id_to_name:
            del self.file_id_to_name[file_id]
            print(f"🗑️  FILE SYNC: Removed {file_id} from filename mapping")

File: monitoring-client-rust/src/modules/file_sync_manager.rs
//! File synchronization manager module
//!
//! Manages OTA (Over-The-Air) file transfers from server to client.
//! Supports parallel downloads, status tracking, and file deletion synchronization.
//!
//! # Features
//! - Server polling for pending files
//! - Parallel downloads (max 5 concurrent)
//! - Download status tracking and reporting
//! - File deletion synchronization
//! - Automatic retry on failures
//! - Thread-safe operation
//!
//! # Example
//! ```no_run
//! use monitoring_client::modules::file_sync_manager::FileSyncManager;
//! use monitoring_client::modules::config::Config;
//! use std::sync::Arc;
//! use parking_lot::RwLock;
//!
//! #[tokio::main]
//! async fn main() -> Result<(), Box<dyn std::error::Error>> {
//!     let config = Arc::new(RwLock::new(Config::load(None)?));
//!     let manager = FileSyncManager::new(
//!         config,
//!         "employee@example.com".to_string(),
//!         "client-123".to_string(),
//!     )?;
//!     
//!     // Start file sync
//!     manager.start().await?;
//!     
//!     // Check for files
//!     manager.check_and_download().await?;
//!     
//!     Ok(())
//! }
//! ```

use crate::modules::config::Config;
use crate::modules::error::{MonitoringError, Result};
use parking_lot::RwLock;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::path::PathBuf;
use std::sync::Arc;
use tokio::fs;
use tokio::sync::Semaphore;
use tracing::{debug, error, info, warn};

/// Maximum number of parallel downloads
pub const MAX_PARALLEL_DOWNLOADS: usize = 5;

/// File information from server
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileInfo {
    /// Unique file ID
    pub id: String,
    
    /// Filename
    pub filename: String,
    
    /// File size in bytes
    #[serde(rename = "fileSize")]
    pub file_size: Option<u64>,
}

/// Download status
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum DownloadStatus {
    /// Download is pending
    Pending,
    
    /// Download is in progress
    Downloading,
    
    /// Download completed successfully
    Completed,
    
    /// Download failed
    Failed,
}

/// File sync manager for OTA file transfers
pub struct FileSyncManager {
    /// Shared configuration
    config: Arc<RwLock<Config>>,
    
    /// Employee name
    employee_name: String,
    
    /// Client ID
    client_id: String,
    
    /// Download directory path
    download_path: PathBuf,
    
    /// HTTP client
    client: Client,
    
    /// Set of downloaded file IDs
    downloaded_files: Arc<RwLock<HashSet<String>>>,
    
    /// Map of file ID to filename
    file_id_to_name: Arc<RwLock<HashMap<String, String>>>,
    
    /// Semaphore for limiting parallel downloads
    download_semaphore: Arc<Semaphore>,
    
    /// Check counter for logging
    check_counter: Arc<RwLock<u32>>,
    
    /// Running flag
    running: Arc<RwLock<bool>>,
}

impl FileSyncManager {
    /// Create a new file sync manager
    ///
    /// # Arguments
    /// * `config` - Shared configuration
    /// * `employee_name` - Employee name
    /// * `client_id` - Unique client ID
    ///
    /// # Returns
    /// A new FileSyncManager instance
    pub fn new(
        config: Arc<RwLock<Config>>,
        employee_name: String,
        client_id: String,
    ) -> Result<Self> {
        let download_path = {
            let cfg = config.read();
            cfg.file_download_path.clone()
        };
        
        let client = Client::builder()
            .timeout(std::time::Duration::from_secs(300)) // 5 minutes for large files
            .build()
            .map_err(|e| MonitoringError::FileSync(format!("Failed to create HTTP client: {}", e)))?;
        
        Ok(Self {
            config,
            employee_name,
            client_id,
            download_path,
            client,
            downloaded_files: Arc::new(RwLock::new(HashSet::new())),
            file_id_to_name: Arc::new(RwLock::new(HashMap::new())),
            download_semaphore: Arc::new(Semaphore::new(MAX_PARALLEL_DOWNLOADS)),
            check_counter: Arc::new(RwLock::new(0)),
            running: Arc::new(RwLock::new(false)),
        })
    }
    
    /// Start the file sync manager
    pub async fn start(&self) -> Result<()> {
        if *self.running.read() {
            warn!("⚠️ File sync manager already running");
            return Ok(());
        }
        
        *self.running.write() = true;
        
        // Ensure download directory exists
        fs::create_dir_all(&self.download_path)
            .await
            .map_err(|e| MonitoringError::FileSync(format!("Failed to create download directory: {}", e)))?;
        
        info!("📁 File sync manager initialized. Download path: {:?}", self.download_path);
        info!("🆔 Client ID: {}", self.client_id);
        info!("✅ File sync manager started");
        
        Ok(())
    }
    
    /// Stop the file sync manager
    pub fn stop(&self) {
        *self.running.write() = false;
        info!("🛑 File sync manager stopped");
    }
    
    /// Update file sync - check for files
    ///
    /// This is called periodically from the monitoring loop
    pub async fn update(&self) -> Result<()> {
        if !*self.running.read() {
            return Ok(());
        }
        
        *self.check_counter.write() += 1;
        
        self.check_and_download().await
    }
    
    /// Force an immediate check for pending files and process downloads
    pub async fn check_and_download(&self) -> Result<()> {
        if !*self.running.read() {
            return Ok(());
        }
        
        info!("🔍 File sync: Force check triggered");
        
        self.check_pending_files().await?;
        
        Ok(())
    }
    
    /// Check server for pending files and sync deletions
    async fn check_pending_files(&self) -> Result<()> {
        let (base_url, auth_token) = {
            let config = self.config.read();
            let base_url = config.server_url.trim_end_matches("/api/monitoring/data").to_string();
            (base_url, config.auth_token.clone())
        };
        
        // Check for pending files
        let pending_url = format!(
            "{}/api/files/pending/{}?client_id={}",
            base_url, self.employee_name, self.client_id
        );
        
        debug!("🌐 Checking pending files at: {}", pending_url);
        
        let response = self
            .client
            .get(&pending_url)
            .header("Authorization", format!("Bearer {}", auth_token))
            .send()
            .await
            .map_err(|e| MonitoringError::FileSync(format!("Failed to check pending files: {}", e)))?;
        
        if response.status().is_success() {
            #[derive(Deserialize)]
            struct PendingFilesResponse {
                files: Vec<FileInfo>,
            }
            
            let data: PendingFilesResponse = response
                .json()
                .await
                .map_err(|e| MonitoringError::FileSync(format!("Failed to parse response: {}", e)))?;
            
            let files = data.files;
            
            if !files.is_empty() {
                info!("📦 Found {} pending file(s)", files.len());
            } else {
                let counter = *self.check_counter.read();
                if counter % 6 == 1 {
                    debug!("✅ No pending files (check #{})", counter);
                }
            }
            
            // Download files in parallel
            let mut tasks = Vec::new();
            
            for file_info in files {
                let file_id = file_info.id.clone();
                let filename = file_info.filename.clone();
                
                // Check if already downloaded
                if self.downloaded_files.read().contains(&file_id) {
                    debug!("⏭️ File already downloaded: {}", filename);
                    continue;
                }
                
                info!("📄 File: {} (ID: {}, Size: {:?})", filename, file_id, file_info.file_size);
                info!("✅ Queued for download: {}", filename);
                
                // Spawn download task
                let task = self.download_file(file_info);
                tasks.push(task);
            }
            
            // Wait for all downloads to complete
            for task in tasks {
                let _ = task.await;
            }
        } else {
            warn!("⚠️ Failed to check pending files: HTTP {}", response.status());
        }
        
        // Check for deleted files
        debug!("🔍 Checking for deleted files...");
        self.sync_deletions(&base_url, &auth_token).await?;
        
        Ok(())
    }
    
    /// Download a file from the server
    async fn download_file(&self, file_info: FileInfo) -> Result<()> {
        let file_id = file_info.id.clone();
        let filename = file_info.filename.clone();
        
        // Acquire semaphore permit for parallel download limiting
        let _permit = self.download_semaphore.acquire().await.map_err(|e| {
            MonitoringError::FileSync(format!("Failed to acquire download permit: {}", e))
        })?;
        
        info!("⬇️ Starting download: {}", filename);
        
        // Update status to downloading
        self.update_status(&file_id, DownloadStatus::Downloading, None).await?;
        
        let (base_url, auth_token) = {
            let config = self.config.read();
            let base_url = config.server_url.trim_end_matches("/api/monitoring/data").to_string();
            (base_url, config.auth_token.clone())
        };
        
        // Download file
        let url = format!("{}/api/files/{}/download", base_url, file_id);
        
        debug!("🌐 Download URL: {}", url);
        
        let response = self
            .client
            .get(&url)
            .header("Authorization", format!("Bearer {}", auth_token))
            .query(&[("employeeName", &self.employee_name)])
            .send()
            .await;
        
        match response {
            Ok(resp) if resp.status().is_success() => {
                // Save file
                let file_path = self.download_path.join(&filename);
                
                info!("💾 Saving file to: {:?}", file_path);
                
                let bytes = resp
                    .bytes()
                    .await
                    .map_err(|e| MonitoringError::FileSync(format!("Failed to read response: {}", e)))?;
                
                fs::write(&file_path, &bytes)
                    .await
                    .map_err(|e| MonitoringError::FileSync(format!("Failed to write file: {}", e)))?;
                
                info!("✅ Download completed: {} ({} bytes)", filename, bytes.len());
                
                // Update status to completed
                self.update_status(&file_id, DownloadStatus::Completed, None).await?;
                
                // Track downloaded file
                self.downloaded_files.write().insert(file_id.clone());
                self.file_id_to_name.write().insert(file_id, filename);
                
                Ok(())
            }
            Ok(resp) => {
                let status = resp.status();
                let error_msg = format!("Download failed with status {}", status);
                error!("❌ {}: {}", error_msg, filename);
                
                self.update_status(&file_id, DownloadStatus::Failed, Some(&error_msg)).await?;
                
                Err(MonitoringError::FileSync(error_msg))
            }
            Err(e) => {
                let error_msg = format!("Download request failed: {}", e);
                error!("❌ Error downloading {}: {}", filename, error_msg);
                
                self.update_status(&file_id, DownloadStatus::Failed, Some(&error_msg)).await?;
                
                Err(MonitoringError::FileSync(error_msg))
            }
        }
    }
    
    /// Update download status on server
    async fn update_status(
        &self,
        file_id: &str,
        status: DownloadStatus,
        error: Option<&str>,
    ) -> Result<()> {
        let (base_url, auth_token) = {
            let config = self.config.read();
            let base_url = config.server_url.trim_end_matches("/api/monitoring/data").to_string();
            (base_url, config.auth_token.clone())
        };
        
        let url = format!("{}/api/files/{}/status", base_url, file_id);
        
        #[derive(Serialize)]
        struct StatusUpdate<'a> {
            #[serde(rename = "employeeName")]
            employee_name: &'a str,
            status: DownloadStatus,
            #[serde(skip_serializing_if = "Option::is_none")]
            error: Option<&'a str>,
        }
        
        let data = StatusUpdate {
            employee_name: &self.employee_name,
            status,
            error,
        };
        
        let response = self
            .client
            .post(&url)
            .header("Authorization", format!("Bearer {}", auth_token))
            .header("Content-Type", "application/json")
            .json(&data)
            .send()
            .await;
        
        match response {
            Ok(resp) if !resp.status().is_success() => {
                warn!("Failed to update status: {}", resp.status());
            }
            Err(e) => {
                error!("Error updating status: {}", e);
            }
            _ => {}
        }
        
        Ok(())
    }
    
    /// Check for deleted files and remove them from local storage
    async fn sync_deletions(&self, base_url: &str, auth_token: &str) -> Result<()> {
        let active_url = format!("{}/api/files/active/{}", base_url, self.employee_name);
        
        debug!("🔍 Checking for deleted files at: {}", active_url);
        
        let response = self
            .client
            .get(&active_url)
            .header("Authorization", format!("Bearer {}", auth_token))
            .send()
            .await
            .map_err(|e| MonitoringError::FileSync(format!("Failed to check active files: {}", e)))?;
        
        if response.status().is_success() {
            #[derive(Deserialize)]
            struct ActiveFilesResponse {
                files: Vec<FileInfo>,
            }
            
            let data: ActiveFilesResponse = response
                .json()
                .await
                .map_err(|e| MonitoringError::FileSync(format!("Failed to parse response: {}", e)))?;
            
            // Build set of filenames that should exist
            let server_filenames: HashSet<String> = data.files.iter().map(|f| f.filename.clone()).collect();
            
            debug!("📋 Server has {} active file(s)", server_filenames.len());
            
            // Check local files
            if self.download_path.exists() {
                let mut entries = fs::read_dir(&self.download_path)
                    .await
                    .map_err(|e| MonitoringError::FileSync(format!("Failed to read directory: {}", e)))?;
                
                while let Some(entry) = entries
                    .next_entry()
                    .await
                    .map_err(|e| MonitoringError::FileSync(format!("Failed to read entry: {}", e)))?
                {
                    let path = entry.path();
                    
                    if path.is_file() {
                        if let Some(filename) = path.file_name().and_then(|n| n.to_str()) {
                            // If file exists locally but not on server, delete it
                            if !server_filenames.contains(filename) {
                                match fs::remove_file(&path).await {
                                    Ok(_) => {
                                        info!("🗑️ Deleted orphaned file: {}", filename);
                                        
                                        // Remove from tracking
                                        let mut file_id_to_remove = None;
                                        {
                                            let mapping = self.file_id_to_name.read();
                                            for (fid, fname) in mapping.iter() {
                                                if fname == filename {
                                                    file_id_to_remove = Some(fid.clone());
                                                    break;
                                                }
                                            }
                                        }
                                        
                                        if let Some(fid) = file_id_to_remove {
                                            self.downloaded_files.write().remove(&fid);
                                            self.file_id_to_name.write().remove(&fid);
                                        }
                                    }
                                    Err(e) => {
                                        error!("❌ Error deleting {}: {}", filename, e);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        } else {
            warn!("⚠️ Could not check for deletions: HTTP {}", response.status());
        }
        
        Ok(())
    }
    
    /// Handle file uploaded event (e.g., from WebSocket)
    pub async fn handle_file_uploaded(&self, filename: &str) -> Result<()> {
        info!("📢 Received file upload notification: {}", filename);
        self.check_and_download().await
    }
    
    /// Handle file deleted event (e.g., from WebSocket)
    pub async fn handle_file_deleted(&self, file_id: &str, filename: Option<&str>) -> Result<()> {
        info!("🗑️ Received file deletion notification: {}", file_id);
        
        // Try to get filename from mapping or use provided filename
        let fname = {
            let mapping = self.file_id_to_name.read();
            mapping.get(file_id).cloned().or_else(|| filename.map(|s| s.to_string()))
        };
        
        if let Some(fname) = fname {
            let file_path = self.download_path.join(&fname);
            
            match fs::remove_file(&file_path).await {
                Ok(_) => {
                    info!("✅ Deleted local file: {}", fname);
                }
                Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
                    warn!("⚠️ File not found locally: {}", fname);
                }
                Err(e) => {
                    error!("❌ Error deleting file {}: {}", fname, e);
                    return Err(MonitoringError::FileSync(format!("Failed to delete file: {}", e)));
                }
            }
        } else {
            error!("❌ Cannot delete - filename unknown for ID: {}", file_id);
        }
        
        // Remove from tracking
        self.downloaded_files.write().remove(file_id);
        self.file_id_to_name.write().remove(file_id);
        
        Ok(())
    }
    
    /// Get the download path
    pub fn download_path(&self) -> &PathBuf {
        &self.download_path
    }
    
    /// Check if running
    pub fn is_running(&self) -> bool {
        *self.running.read()
    }
    
    /// Get number of downloaded files
    pub fn downloaded_count(&self) -> usize {
        self.downloaded_files.read().len()
    }
}
